WordPress/gutenberg
# Allow customising the Navigation block breakpoint

## What

Adds an `overlayBreakpoint` attribute to the Navigation block so the responsive overlay can switch between the collapsed menu button and inline navigation items at a user-configured viewport width.

This is a major refresh of the earlier PR direction. The PR had gone stale, so the implementation has been updated to fit the current Navigation block overlay architecture and the newer patterns now present in `trunk`.

## Why

The Navigation block currently uses a fixed breakpoint for its mobile overlay behavior. That can be too limiting for themes and layouts where navigation needs to collapse earlier or later depending on available space, typography, item count, or design requirements.

## How

- Adds an `overlayBreakpoint` block attribute with a default of `600px`.
- Adds an Overlay breakpoint control to the Navigation block Overlay settings when Overlay Visibility is set to Mobile.
- Supports `px`, `em`, and `rem` values, including decimal values.
- Normalizes invalid or unsafe values back to the default breakpoint.
- Preserves the existing default breakpoint behavior for Navigation blocks without a custom value.
- Emits instance-scoped frontend CSS only when a valid custom mobile overlay breakpoint is used.
- Mirrors the custom breakpoint behavior in the editor preview.
- Ignores custom breakpoint styling for Overlay Visibility values that do not use a responsive breakpoint, such as Off and Always.
- Updates Navigation block docs and adds JS/PHP coverage for validation, editor behavior, rendered CSS, and default compatibility.

## FAQs

### Why 600px?

- `600px` is the breakpoint Navigation already uses today through WordPress's small breakpoint, not a new design decision introduced by this PR.
- `packages/base-styles/_breakpoints.scss:12` defines `$break-small: 600px`.
- `packages/base-styles/_mixins.scss:118` exposes that value as `break-small()`.
- Keeping that value as the default preserves the current behavior for existing Navigation blocks.
- Custom values only apply when a block has a valid non-default Mobile overlay breakpoint, and those styles are scoped to the individual Navigation block instance.

### Why not automate this?

Two automation routes were explored before returning to an explicit breakpoint control:

- CSS/intrinsic layout, including container queries.
- JavaScript measurement, where the block detects at runtime whether the navigation still fits.

Those routes did not give us a robust enough solution for this PR:

- Container queries were suggested in [#45274](https://github.com/WordPress/gutenberg/pull/45274#issuecomment-1292037460) and explored further in [#53475](https://github.com/WordPress/gutenberg/discussions/53475).
- The blocker is that container queries still require an explicit threshold, for example "when this container is narrower than `X`, switch layout." They do not automatically know when the Navigation block's inline items no longer fit.
- The correct `X` is content-dependent. It changes with the number of menu items, label length, font size, spacing, submenu indicators, and translations. A threshold that works for one Navigation instance can be wrong for another instance in the same theme.
- Querying the Navigation block's own container is not enough because many real headers break due to surrounding layout, not only the menu itself. Site Logo, Site Title, social links, Groups, Columns, spacing, and future layout containers can all reduce the available space.
- Querying a parent container does not solve that either. It just moves the hardcoded threshold to a different container, and the right parent can vary across header layouts.
- [#57587](https://github.com/WordPress/gutenberg/pull/57587) explored a JavaScript `auto` mode because CSS/container queries could not answer the actual question: "do these navigation items still fit in this layout right now?" That route required measurement heuristics across both Navigation items and parent containers.
- Those heuristics surfaced unresolved concerns around layout assumptions, nested containers, zooming, multiple collapsible menus, resize handling, editor-visible wrapped states, measurement performance, and future Grid layout changes.
- This PR therefore takes the narrower route: preserve the default behavior, expose a custom value only for the Mobile overlay case, and scope generated CSS to the individual Navigation block instance.

### How does this relate to the WordPress 7.1 responsive styling system?

- The new responsive styling system and this PR solve different problems.
- The WordPress 7.1 responsive styling system lets block styles vary across named viewport states such as `@mobile` and `@tablet`. Those states are driven by `settings.viewport.mobile` and `settings.viewport.tablet`.
- `overlayBreakpoint` is not a responsive style state. It is a Navigation-block behavior setting that controls when one Navigation block switches between inline items and the overlay menu button.
- Changing `settings.viewport` does not automatically change a Navigation block's `overlayBreakpoint`, and changing `overlayBreakpoint` does not redefine the site's `@mobile` or `@tablet` style breakpoints.
- The two systems can still work together. Responsive style controls can adjust how the Navigation block looks at mobile/tablet viewports, while `overlayBreakpoint` controls when the overlay behavior starts.
- In practice, a theme or user can choose matching values if they want the Navigation overlay transition to align with a site viewport breakpoint, but this PR does not force that coupling.
- Keeping the values separate avoids making a content-dependent Navigation behavior part of the global responsive styling API, while still preserving compatibility with the new responsive styling system.

## Testing Instructions

1. Add a Navigation block.
2. Open the block settings sidebar and expand the Overlay panel.
3. Set Overlay Visibility to Mobile.
4. Change the Overlay breakpoint value, for example `720px`, `37.5em`, or `48rem`.
5. Resize the editor and frontend viewport and confirm the Navigation block switches between the menu button and inline navigation items at the configured breakpoint.
6. Confirm a Navigation block using the default `600px` breakpoint still behaves as before.
7. Confirm the Overlay breakpoint control is hidden when Overlay Visibility is Off or Always.
8. Run `npm run test:unit packages/block-library/src/navigation/edit/test`.
9. With the test environment running, run `vendor/bin/phpunit phpunit/blocks/render-block-navigation-test.php`.
